### DISCUSSION TITLE
E2 Agent Support for latest FlexRIC Commit to deploy E2 RAN Control (RC) xApps

### DISCUSSION THREAD

**Discussion by RobinWie** [srsran/srsRAN_Project#409] [General Help]
Hello everyone,

I have successfully completed the official O-RAN tutorial [[O-RAN Tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html)](https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html) for deploying and testing an E2 KPM xApp in combination with FlexRIC and the srsRAN E2 agent. Following the exact software versions and commits specified in the tutorial, this setup works for me without issues.

I am currently developing a custom xApp that requires the integration of E2 RAN Control (RC) functionality in addition to the KPM service model outlined in the tutorial. According to the tutorial, srsRAN supports the O-RAN.WG3.E2SM-RC-R003-v03.00 Ran Control service model. However, the current E2 Agent seems to only be compatible with an older branch of FlexRIC (e2ap-v2), which, to my understanding, does not support any RAN Control service models.

To address this, I have tried using the latest commits of both srsRAN and FlexRIC (current FlexRIC commit: 70dd779f5339f4dac3ac176a24b49c45046a59fa), which includes E2 RAN Control v1.03. My goal was to successfully send Control Requests from the RC xApp to srsRAN using the kpm_rc xapp. Unfortunately, it appears that srsRAN is unable to decode the Control Requests as shown by the logs below.

Has anyone encountered a similar issue and found a viable workaround? Additionally, is there any chance that srsRAN will extend support to more recent versions of FlexRIC in the near future?

Thank you for your assistance!
[flexric_log.txt](https://github.com/srsran/srsRAN_Project/files/13963813/flexric_log.txt)
[gnb_log.txt](https://github.com/srsran/srsRAN_Project/files/13963814/gnb_log.txt)
[xapp_kpm_rc_log.txt](https://github.com/srsran/srsRAN_Project/files/13963815/xapp_kpm_rc_log.txt)

**Comment by yagoda** [comment]
Hi Robin!
Can you do the following:
1. take the gnb.log from /tmp and share it here (make sure you set the log level to info in the config first)
2. take a pcap of the interface between the gNodeB and the FlexRic (tcpdump -i <interface> sctp -w save_e2.pcap)

(We will soon have an updated app note for the latest version of FlexRic)

Regards,
Justin

**Reply by RobinWie** [reply]
Hi Justin,
thanks for your reply! I just updated to the latest source code and now it seems that the problems lies somewhere in the xApp since srsRAN is able to send a E2 setup response. I append the requested logs. I captured the e2 traffic on the loopback interface as suggested in the tutorial.

Regards
Robin

[flexric.txt](https://github.com/srsran/srsRAN_Project/files/13991817/flexric.txt)
[xapp.txt](https://github.com/srsran/srsRAN_Project/files/13991820/xapp.txt)
[gnb.log](https://github.com/srsran/srsRAN_Project/files/13992006/gnb.log)
[e2ap_packets.zip](https://github.com/srsran/srsRAN_Project/files/13992013/e2ap_packets.zip)

**Comment by pgawlowicz** [comment]
Hi @RobinWie, 

we now provide an e2sm_rc example with ORAN SC RIC. Please check here: https://github.com/srsran/oran-sc-ric

**Comment by AlejandraOliver** [comment]
Hi @RobinWie , I need to create a custom xApp and I dont know how to start. Could yo give some steps in order to build it? I need for a big proyect and I’m in hurry!! Thanks!!!

**Reply by newt2105** [reply]
Hi @AlejandraOliver, Have you created new xapp or custom xapp yet?
